tiatoolbox/tools/pyramid.py

```diff
@@ -0,0 +1,526 @@
+        path = Path(path)
+        if container == "zip":
+            compression2enum = {
+                None: zipfile.ZIP_STORED,
+                "deflate": zipfile.ZIP_DEFLATED,
+                "bz2": zipfile.ZIP_BZIP2,
+                "lzma": zipfile.ZIP_LZMA,
+            }
+            if compression not in compression2enum:
+                raise ValueError("Unsupported compression for zip")
+            archive = zipfile.ZipFile(
+                path, mode="w", compression=compression2enum[compression]
+            )
+        elif container == "tar":
+            compression2mode = {
+                None: "w",
+                "gzip": "w:gz",
+                "bz2": "w:bz2",
+                "lzma": "w:xz",
+            }
+            if compression not in compression2mode:
+                raise ValueError("Unsupported compression for tar")
+
+            archive = tarfile.TarFile.open(path, mode=compression2mode[compression])
+        elif container is None:
+            path.mkdir(parents=False)
+            if compression is not None:
+                raise ValueError("Unsupported compression for container None")
+        else:
+            raise ValueError("Unsupported container")
+        for level in range(self.level_count):
+            for x, y in np.ndindex(self.tile_grid_size(level)):
+                tile = self.get_tile(level=level, x=x, y=y)
+                tile_path = self.tile_path(level, x, y)
+                full_path = path / tile_path
+                if container is None:
+                    full_path.parent.mkdir(parents=True, exist_ok=True)
+                if container is None:
+                    tile.save(full_path)
+                else:
+                    bio = BytesIO()
+                    tile.save(bio, format="jpeg")
+                    bio.seek(0)
+                    if container == "tar":
+                        tar_info = tarfile.TarInfo(name=str(tile_path))
+                        tar_info.mtime = time.time()
+                        tar_info.size = bio.tell()
+                        archive.addfile(tarinfo=tar_info, fileobj=bio)
+                    if container == "zip":
+                        data = bio.read()
+                        archive.writestr(
+                            str(tile_path),
+                            data,
+                            compress_type=compression2enum[compression],
+                        )
+        if container is not None:
+            archive.close()
+
+    def __len__(self) -> int:
+        return sum(
+            np.prod(self.tile_grid_size(level)) for level in range(self.level_count)
+        )
+
+    def __iter__(self) -> Iterable:
+        for level in range(self.level_count):
+            for x, y in np.ndindex(self.tile_grid_size(level)):
+                yield self.get_tile(level=level, x=x, y=y)
+
+
+class DeepZoomGenerator(TilePyramidGenerator):
+    r"""Pyramid tile generator following the DeepZoom format.
+
+    Args:
```

Comment:
Attributes?

Reply:
This would just be an exact duplicate of the argument. Do you really want that?